pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,462 @@
+	preference             string
+	volumePvc              []string
+	volumeDv               []string
+	volumeDvtPvc           []string
+	volumeDvtDs            []string
+	cloudInitUserData      string
+	cloudInitNetworkData   string
+}
+
+type optionFn func(*createVM, *v1.VirtualMachine) error
+
+var optFns = map[string]optionFn{
+	RunStrategyFlag:          withRunStrategy,
+	InstancetypeFlag:         withInstancetype,
+	PreferenceFlag:           withPreference,
+	VolumeCloneDsFlag:        withVolumeDvtDs,
+	VolumeClonePvcFlag:       withVolumeDvtPvc,
+	VolumeDvFlag:             withVolumeDv,
+	VolumePvcFlag:            withVolumePvc,
+	CloudInitUserDataFlag:    withCloudInitData,
+	CloudInitNetworkDataFlag: withCloudInitData,
+}
+
+var flags = []string{
+	RunStrategyFlag,
+	InstancetypeFlag,
+	PreferenceFlag,
+	VolumeCloneDsFlag,
+	VolumeClonePvcFlag,
+	VolumeDvFlag,
+	VolumePvcFlag,
+	CloudInitUserDataFlag,
+	CloudInitNetworkDataFlag,
+}
+
+var runStrategies = map[string]struct{}{
+	string(v1.RunStrategyAlways):         {},
+	string(v1.RunStrategyManual):         {},
+	string(v1.RunStrategyHalted):         {},
+	string(v1.RunStrategyOnce):           {},
+	string(v1.RunStrategyRerunOnFailure): {},
+}
+
+func aliasNormalizeFunc(_ *pflag.FlagSet, flagName string) pflag.NormalizedName {
+	switch flagName {
+	case VolumeDvtPvcFlag:
+		flagName = VolumeClonePvcFlag
+	case VolumeDvtDsFlag:
+		flagName = VolumeCloneDsFlag
+	}
+	return pflag.NormalizedName(flagName)
+}
+
+func NewVirtualMachineCommand() *cobra.Command {
+	c := defaultCreateVM()
+	cmd := &cobra.Command{
+		Use:     VM,
+		Short:   "Create a VirtualMachine manifest.",
+		Example: c.usage(),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return c.run(cmd)
+		},
+	}
+
+	cmd.Flags().StringVar(&c.name, NameFlag, c.name, "Specify the name of the VM")
+	cmd.Flags().Int64Var(&c.terminationGracePeriod, TerminationGracePeriodFlag, c.terminationGracePeriod, "Specify the termination grace period of the VM")
+
+	cmd.Flags().BoolVar(&c.running, RunningFlag, c.running, "Specify if the VM should be running")
+	cmd.Flags().StringVar(&c.runStrategy, RunStrategyFlag, c.runStrategy, "Specify the RunStrategy of the VM")
+	cmd.MarkFlagsMutuallyExclusive(RunningFlag, RunStrategyFlag)
+
+	cmd.Flags().StringVar(&c.instancetype, InstancetypeFlag, c.instancetype, "Specify the Instance Type of the VM")
+	cmd.Flags().StringVar(&c.preference, PreferenceFlag, c.preference, "Specify the Preference of the VM")
+
+	cmd.Flags().StringSliceVar(&c.volumeDvtPvc, VolumeClonePvcFlag, c.volumeDvtPvc, "Specify one or more PVCs to be cloned by the VM")
+	cmd.Flags().StringSliceVar(&c.volumeDvtDs, VolumeCloneDsFlag, c.volumeDvtDs, "Specify one or more DSs to be cloned by the VM")
+	cmd.Flags().StringSliceVar(&c.volumePvc, VolumePvcFlag, c.volumePvc, "Specify one or more PVCs to be used by the VM")
+	cmd.Flags().StringSliceVar(&c.volumeDv, VolumeDvFlag, c.volumeDv, "Specify one or more DVs to be used by the VM")
```

Comment:
I think there's some conflicting views here on the CLI args being used. I'll try and clear some things up.

DataVolumes only exist as a workflow now. There's no reason to assign a DataVolume directly to a VM. Assign the DV's PVC instead. This is partially why I initially recommended something like the ```--image-source``` field, but I i can see that the ```image-source``` are is not expressive enough now.

There are a lot of dimensions here to volumes. There's the source, the storageclass, the volume size, the access modes, disk bus type, etc... I'd like to see a cli arg that can express all that while performing smart defaulting.

Maybe we could make a ```--volume``` argument that takes in multiple options.

For example:

```
// add a volume to the VM. the importSource option indicates the volume is a DVT that is cloning/importing
--volume="name:myvol,importSource:docker://kubevirt/fedora:latest,size:20Gi,bus:virtio,storageClass:ceph"

// add a volume to a VM with an importSource of type container disk.
--volume="name:myvol,importSource:pvc://mysourcepvc,size:20Gi,bus:virtio,storageClass:ceph"

// add a volume to a VM with an importSource of type http
--volume="name:myvol,importSource:http://example.com/mysourcepvc.qcow2,size:20Gi,bus:virtio"

// Add an existing PVC to VM directly 
--volume="name:myvol,pvc:mysourcepvc,bus:virtio"
```

I didn't put a lot of thought into the above examples. The point I'm trying to make here is that we need our cli args to model our most common user workflows and be expressive enough to be useful.

Reply:
> I think there's some conflicting views here on the CLI args being used. I'll try and clear some things up.
> 
> DataVolumes only exist as a workflow now. There's no reason to assign a DataVolume directly to a VM. Assign the DV's PVC instead. This is partially why I initially recommended something like the `--image-source` field, but I i can see that the `image-source` are is not expressive enough now.

That's fair, I was for including it as it is possible through the API at the moment, maybe we should reject or just warn against it when a DVT isn't present and direct users to just point at a PVC instead?

> There are a lot of dimensions here to volumes. There's the source, the storageclass, the volume size, the access modes, disk bus type, etc... I'd like to see a cli arg that can express all that while performing smart defaulting.
>
> Maybe we could make a `--volume` argument that takes in multiple options.
> 
> For example:
> 
> ```
> // add a volume to the VM. the importSource option indicates the volume is a DVT that is cloning/importing
> --volume="name:myvol,importSource:docker://kubevirt/fedora:latest,size:20Gi,bus:virtio,storageClass:ceph"
> 
> // add a volume to a VM with an importSource of type container disk.
> --volume="name:myvol,importSource:pvc://mysourcepvc,size:20Gi,bus:virtio,storageClass:ceph"
> 
> // add a volume to a VM with an importSource of type http
> --volume="name:myvol,importSource:http://example.com/mysourcepvc.qcow2,size:20Gi,bus:virtio"
> 
> // Add an existing PVC to VM directly 
> --volume="name:myvol,pvc:mysourcepvc,bus:virtio"
> ```
> 
> I didn't put a lot of thought into the above examples. The point I'm trying to make here is that we need our cli args to model our most common user workflows and be expressive enough to be useful.

I really wanted to avoid the `encode:the,entire_API:in,a:string` approach, we tried this in both the `novaclient` and `openstackclient` OpenStack CLIs for `block-device-mappings` and it was widely hated by devs and end users alike. I'd much rather see well defined independent switches per `VolumeSource` with limited to no options (to start) as we've tried to model here for the PVC based sources.

For options like `diskBus` and `storageClass` we can model these as preferences _today_ so I'd rather avoid listing these at all. If a user really wants control of these per individual volume or disk then they can always modify the generated manifest before submission. 

For everything else I'd rather get feedback from end users on what they need to control before we add things.

Reply:
> I really wanted to avoid the encode:the,entire_API:in,a:string approach,
And if we base64 encode it? :)

More seriously: I agree with Lee that the virutal ABI parts are in the Pref, and I'd not specify them (initially) on the cmdline. On the cmdline I'd focus on providing the resource which is mandatory: without a vol, a vm wil not boot (ignoring pxe).

Can we start opinionated i.e. with
```
# Direct image use, alices-fedora will be directly attached and used (maps to a plain volume in the VM spec)
$ virtctl create vm alice --volume alices-fedora

# Cloning an image from a shared namespace, aka "golden image" (maps to a DVT with DS in the VM spec)
$ virtctl create vm bob --datasource shared-images/fedora
```

IMHO we could also consider, but this is a longer path
```
# Direct image use - as above
$ virtctl create vm alice --volume alices-fedora

# Cloning an image - as above, different implementation
# … as an admin, prepare:
$ virtctl create volume-populator images/fedora $SOME_ARGS_RELEVANT_FOR_CDI
$ virtctl create volume-populator images/fedora $SOME_ARGS_RELEVANT_FOR_CDI

# … as a user, consume:
$ virtctl create vm bob --volume $BOBSPVC --populate-with images/fedora
```
I know this is an extra mile. But to me it would push the complexity on the admin, and the user can focus on consumption.

Reply:
> > I really wanted to avoid the encode:the,entire_API:in,a:string approach,
> > And if we base64 encode it? :)
> 
> More seriously: I agree with Lee that the virtual ABI parts are in the Pref, and I'd not specify them (initially) on the cmdline. On the cmdline I'd focus on providing the resource which is mandatory: without a vol, a vm wil not boot (ignoring pxe).
> 
> Can we start opinionated i.e. with
> 
> ```shell
> # Direct image use, alices-fedora will be directly attached and used (maps to a plain volume in the VM spec)
> $ virtctl create vm alice n1.xlarge --volume alices-fedora

A plain volume being a `PVC` right?

> # Cloning an image from a shared namespace, aka "golden image" (maps to a DVT with DS in the VM spec)
> $ virtctl create vm bob n1.xlarge --datasource shared-images/fedora
> ```

Okay this makes sense.

> IMHO we could also consider [volume-populators](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-populators-and-data-sources), but this is a longer path:

Lets keep this initial proposal to the currently support methods, we can always look into this in future follow ups.

Reply:
I have to agree with @lyarwood. It will get cumbersome very fast if we try to encapsulate the whole API in commandline flags. IMO, `virtctl create vm` should be a quick and easy way to create VMs based on instance types and preferences. If customized VM definitions are needed, the created manifest can always be modified.

+1 for the opinionated `--datasource` and `--volume` flags. Let's keep it simple for now.

When https://github.com/kubevirt/kubevirt/pull/8480 lands `--infer-instancetype` and `--infer-preference` flags (defaulting to true) could be added. The minimal CLI for creating a new VM from an existing boot source would look like this then:

`virtctl create vm --datasource fedora | kubectl create -f -`

Reply:
> When #8480 lands `--infer-instancetype` and `--infer-preference` flags (defaulting to true) could be added. The minimal CLI for creating a new VM from an existing boot source would look like this then:
> 
> `virtctl create vm --datasource fedora | kubectl create -f -`

Cool, `InferFromVolume ` takes the name of a `Volume` FWIW but I assume we'd just pick the first?